specs/experimental/gov-delegation.md

```diff
@@ -0,0 +1,440 @@
+```
+
+#### `_delegate`
+
+Delegates voting power from the `_delegator` to the specified delegates in `_newDelegations`. This function MUST
+override any existing delegations of `_delegator`.
+
+```solidity
+function _delegate(address _delegator, Delegation[] memory _newDelegations) internal virtual
+```
+
+The `_delegate` function MUST first check that the length of `_newDelegations` does not exceed `MAX_DELEGATIONS`.
+Then, it MUST calculate the voting power of all active delegations of the `_delegator`, and calculate the voting
+power adjustments for the new delegatee set using `_newDelegations`. Afterwards, the function MUST update
+the voting power of the current and new set of delegatees. Throughout this process, the function MUST use the
+`DelegationAdjustment` and `Op` types to store voting power adjustments and the operation to perform (add or
+subtract voting power).
+
+The function MUST then sort the new delegatee set in descending order, and store the delegations as part of the
+`_delegations` mapping. While sorting, the function MUST check for the validity of the delegations, as specified
+in the [Delegation Validation](#delegation-validation) section. Additionally, the function MUST emit a
+`DelegateVotesChanged` event for each change in delegatee voting power.
+
+### Getters
+
+For backwards compatibility, the `GovernanceDelegation` contract MUST implement all public getter functions of the
+`GovernanceToken` related to delegation and voting power. These functions MUST be used by the
+`GovernanceToken` when an account has been migrated to the `GovernanceDelegation` contract. Otherwise,
+the `GovernanceToken` MUST use its own state. Similarly, all of the `GovernanceDelegation` getter functions
+MUST use the `GovernanceToken` state if the account has not been migrated.
+
+#### `checkpoints`
+
+Retrieves the checkpoints for a given user address.
+
+```solidity
+function checkpoints(address _account) external view returns (ERC20Votes.Checkpoint[] memory)
+```
+
+#### `numCheckpoints`
+
+Retrieves the number checkpoints for a given user address.
+
+```solidity
+function numCheckpoints(address _account) external view returns (uint32)
+```
+
+#### `delegates`
+
+Retrieves the delegation of a given user address with the highest voting power.
+This function is intended to be used by the `GovernanceToken` contract to maximize for backwards compatibility.
+
+```solidity
+function delegates(address _account) external view returns (address)
+```
+
+#### `getVotes`
+
+Retrieves the voting power of a given user address.
+
+```solidity
+function getVotes(address _account) external view returns (uint256)
+```
+
+#### `getPastVotes`
+
+Retrieves the voting power of a given user address at a block.
+
+```solidity
+function getPastVotes(address _account, uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `getPastTotalSupply`
+
+Retrieves the total supply of the `GovernanceToken` at a given block.
```

Comment:
Let's specify that it's an L1 block number

Reply:
Why does L1 block number make more sense here? Isn't it an L2 native token?

Reply:
Going to resolve this since it's a discussion best left for the OP interop design doc. Problem statement is that w need to be able to coordinate voting power from different chains which will have different L2 block numbers at the same time. We therefore need to either use timestamp or L1 block number. There are arguments for each. Since this is not relevant to this specific spec, I will close this comment for now.